modules/virtualnetwork/main.tf

```diff
@@ -155,10 +156,96 @@ resource "azapi_resource" "vhubconnection" {
           id = each.value.vwan_associated_routetable_resource_id != "" ? each.value.vwan_associated_routetable_resource_id : "${each.value.vwan_hub_resource_id}/hubRouteTables/defaultRouteTable"
         }
         propagatedRouteTables = {
-          ids    = local.vwan_propagated_routetables_resource_ids[each.key]
-          labels = local.vwan_propagated_routetables_labels[each.key]
+          ids    = each.value.vwan_security_configuration.secure_private_traffic == true ? local.vwan_propagated_noneroutetables_resource_ids[each.key] : local.vwan_propagated_routetables_resource_ids[each.key]
```

Comment:
Please can you explain what this does? You are specifying the `noneRouteTable` in propagatedRouteTables when `secure_private_traffic` is enabled. The default behaviour was to use `defaultRouteTable`.

Why is this necessary when using secure_private_traffic?

Thanks

Reply:
In security configuration when private traffic is set to "Bypass Azure Firewall"
<img width="1009" alt="Screenshot 2023-01-26 at 10 06 00" src="https://user-images.githubusercontent.com/7663736/214812684-5f7a22cc-18cc-49d8-8059-d21158818b64.png">
The connection object looks like this 
```json
{
  "name": "vhc-0e2426dc-d34c-5bb5-8b01-8dc1a999084d",
  "id": "/subscriptions/64db9221-622e-4f35-8c66-0000/resourceGroups/Connectivity/providers/Microsoft.Network/virtualHubs/vhub-uksouth/hubVirtualNetworkConnections/vhc-0e2426dc-d34c-5bb5-8b01-8dc1a999084d",
  "etag": "W/\"bbde0027-4cc5-4599-898f-3b6e415fb4ad\"",
  "type": "Microsoft.Network/virtualHubs/hubVirtualNetworkConnections",
  "properties": {
    "provisioningState": "Succeeded",
    "resourceGuid": "843c14ab-a070-457a-a085-0000",
    "routingConfiguration": {
      "associatedRouteTable": {
        "id": "/subscriptions/64db9221-622e-4f35-8c66-0000/resourceGroups/Connectivity/providers/Microsoft.Network/virtualHubs/vhub-uksouth/hubRouteTables/defaultRouteTable"
      },
      "propagatedRouteTables": {
        "labels": [
          "default"
        ],
        "ids": [
          {
            "id": "/subscriptions/64db9221-622e-4f35-8c66-0000/resourceGroups/Connectivity/providers/Microsoft.Network/virtualHubs/vhub-uksouth/hubRouteTables/defaultRouteTable"
          }
        ]
      },
      "vnetRoutes": {
        "staticRoutes": []
      }
    },
    "remoteVirtualNetwork": {
      "id": "/subscriptions/bb947477-068c-42c6-ac2c-0000/resourceGroups/connectivity-uks/providers/Microsoft.Network/virtualNetworks/vnet_testing_uks"
    },
    "allowHubToRemoteVnetTransit": true,
    "allowRemoteVnetToUseHubVnetGateways": true,
    "enableInternetSecurity": false
  }
}
```
Where `propagatedRouteTables` labels and ids are `default` and `defaultRouteTable` respectively.

When secure private traffic is enabled for the api updates the connection object, it changes.
<img width="1066" alt="Screenshot 2023-01-26 at 10 21 33" src="https://user-images.githubusercontent.com/7663736/214813388-c7efc189-3963-42a6-b6ab-b5e5cda6c15c.png">
and now looks like this
```json
{
  "name": "vhc-0e2426dc-d34c-5bb5-8b01-8dc1a999084d",
  "id": "/subscriptions/64db9221-622e-4f35-8c66-0000/resourceGroups/Connectivity/providers/Microsoft.Network/virtualHubs/vhub-uksouth/hubVirtualNetworkConnections/vhc-0e2426dc-d34c-5bb5-8b01-8dc1a999084d",
  "etag": "W/\"af2e1f66-25a1-439e-9873-56daf33fdcd9\"",
  "type": "Microsoft.Network/virtualHubs/hubVirtualNetworkConnections",
  "properties": {
    "provisioningState": "Updating",
    "resourceGuid": "843c14ab-a070-457a-a085-0000",
    "routingConfiguration": {
      "associatedRouteTable": {
        "id": "/subscriptions/64db9221-622e-4f35-8c66-0000/resourceGroups/Connectivity/providers/Microsoft.Network/virtualHubs/vhub-uksouth/hubRouteTables/defaultRouteTable"
      },
      "propagatedRouteTables": {
        "labels": [
          "none"
        ],
        "ids": [
          {
            "id": "/subscriptions/64db9221-622e-4f35-8c66-0000/resourceGroups/Connectivity/providers/Microsoft.Network/virtualHubs/vhub-uksouth/hubRouteTables/noneRouteTable"
          }
        ]
      },
      "vnetRoutes": {
        "staticRoutes": []
      }
    },
    "remoteVirtualNetwork": {
      "id": "/subscriptions/bb947477-068c-42c6-ac2c-0000/resourceGroups/connectivity-uks/providers/Microsoft.Network/virtualNetworks/vnet_testing_uks"
    },
    "allowHubToRemoteVnetTransit": true,
    "allowRemoteVnetToUseHubVnetGateways": true,
    "enableInternetSecurity": false
  }
```
Where `propagatedRouteTables` labels and ids become `none` and `noneRouteTable` respectively. The code is literally just saying 

```bash
if secure_traffic is true
  labels and ids = none and noneRouteTable
else 
  labels and ids = default and defaultRouteTable
```
To answer more specifically your question as to *why* I'm not sure, I'd speculate that since traffic now has to route via the fw that how the routes are advertised has to change.

Reply:
So propagating to none from what I can see essentially isolates the vnet where its routes aren't published to the hub. I'd asusme this is how the dynamic managed routing of vwan is turned off, where it's only reachable via the fw router and fw policy.

Reply:
Thanks @b0bu, this makes sense. I will double check but think I'm happy with this :) Thanks for your efforts here

Reply:
As an update for anyone that finds this PR, I added the routing config to my secure-vhub module with options which allows me to set the routing globally at the firewall. Now all I need is to get this merged so we can enable it at the connection!  

```hcl
module "secure-vhub-ukwest" {
  ...
  secure_private_traffic = false
  secure_public_traffic  = false
  ...
}
```
where module contents has
```hcl
variable "secure_private_traffic" {
   type = bool
   default = true
}
variable "secure_public_traffic" {
   type = bool
   default = true
}

resource azurerm_virtual_hub_route_table_route "private" {
  for_each = var.secure_private_traffic && !var.secure_public_traffic ? toset(["1"]) : []
  route_table_id = azurerm_virtual_hub.vhub.default_route_table_id

  name             = "private_traffic"
  destination_type = "CIDR"
  destinations     = ["10.0.0.0/8","172.16.0.0/12","192.168.0.0/16"]
  next_hope_type   = "ResourceId"
  next_hop         = azure_firewall.fw.id 
}

resource azurerm_virtual_hub_route_table_route "public" {
  for_each = var.secure_public_traffic && !var.secure_private_traffic ? toset(["1"]) : []
  route_table_id = azurerm_virtual_hub.vhub.default_route_table_id

  name             = "public_traffic"
  destination_type = "CIDR"
  destinations     = ["0.0.0.0/0"]
  next_hope_type   = "ResourceId"
  next_hop         = azure_firewall.fw.id 
}

resource azurerm_virtual_hub_route_table_route "all" {
  for_each = var.secure_private_traffic && var.secure_public_traffic ? toset(["1"]) : []
  route_table_id = azurerm_virtual_hub.vhub.default_route_table_id

  name             = "all_traffic"
  destination_type = "CIDR"
  destinations     = ["0.0.0.0/0", "10.0.0.0/8","172.16.0.0/12","192.168.0.0/16"]
  next_hope_type   = "ResourceId"
  next_hop         = azure_firewall.fw.id 
}
```